exec/cnex/nstring.c

```diff
@@ -208,6 +222,111 @@ size_t string_findChar(TString *self, char c)
     return NPOS;
 }
 
+int64_t string_findFirstOf(TString *self, size_t pos, const char *str)
+{
+    if (self->data == NULL || self->length == 0 || pos > self->length || str == NULL) {
+        return -1;
+    }
+
+    for (size_t l = pos; l < self->length; l++) {
+        if (strchr(str, self->data[l]) != NULL) {
+            return l;
+        }
+    }
+    return -1;
+}
+
+static inline BOOL isNotOneOf(TString *self, char c)
+{
+    for (size_t i = 0; i < self->length; i++) {
+        if (self->data[i] == c) {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+int64_t string_findFirstNotOf(TString *self, TString *p)
+{
+    if (self->data == NULL || self->length == 0 || p == NULL || p->data == NULL) {
+        return -1;
+    }
+
+    // If a string of no chars is passed, should we return 0, or -1?  Technically, we didn't find anything,
+    // so the first char we didn't find would be 0.
+    if (p->length == 0) {
+        return 0;
+    }
+
+    for (size_t l = 0; l < self->length; l++) {
+        if (isNotOneOf(p, self->data[l])) {
+            return l;
+        }
+    }
+    return -1;
+}
+
+int64_t string_findLastNotOf(TString *self, TString *p)
+{
+    if (self->data == NULL || self->length == 0) {
+        return -1;
+    }
+
+    if (p == NULL || p->data == NULL || p->length == 0) {
+        return self->length;
+    }
+
+    for (size_t l = self->length; l > 0; l--) {
+        if (isNotOneOf(p, self->data[l-1])) {
+            return l;
+        }
+    }
+    return -1;
+}
+
+int64_t string_findString(TString *self, size_t pos, TString *p)
+{
+    int64_t r = -1;
+    if (p == NULL || p->data == NULL || p->length == 0) {
+        return r;
+    }
+
+    if (self == NULL || self->data == NULL || self->length == 0 || pos > self->length) {
+        return r;
+    }
+
+    for (size_t l = pos; l < self->length; l++) {
+        if (self->data[l] == p->data[0]) {
+            r = l;
+            if (self->length < (l + p->length)) {
```

Comment:
This check doesn't need to be done here, after finding a first character match. You can do it by calculating the `for` loop bounds for line 298. You only need to search as far as `self->length - p->length`, because any further positions can't possibly hold the target.

Reply:
I don't follow, here.  I have to check the remainder of the string, because the starting char could be anywhere in the string.  When I changed this, it broke my string tests, so I'm trying to figure out what went wrong.  It seems that once we reach the char we're looking for, we break out of the loop.  It seems to make sense, but in practice, it's not working as I'd expect.

Reply:
Fixed logic to only search viable area of string.